app/src/behaviors/behavior_macro.c

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright (c) 2022 The ZMK Contributors
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#define DT_DRV_COMPAT zmk_behavior_macro
+
+#include <device.h>
+#include <drivers/behavior.h>
+#include <logging/log.h>
+#include <zmk/behavior.h>
+#include <zmk/behavior_queue.h>
+#include <zmk/keymap.h>
+
+LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
+
+#if DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT)
+
+enum behavior_macro_mode {
+    MACRO_MODE_TAP,
+    MACRO_MODE_PRESS,
+    MACRO_MODE_RELEASE,
+};
+
+struct behavior_macro_trigger_state {
+    uint32_t wait_ms;
+    uint32_t tap_ms;
+    enum behavior_macro_mode mode;
+    int8_t start_index;
+    int8_t end_index;
```

Comment:
This allows for 127 item macros, which is certainly a lot but still seems within the realm of possibility that someone would run past that limit. Some alternative ideas:

1. Use int16_t instead.
2. Use uint8_t and see my comment below about changing "end" to mean one past the last item.

Reply:
Switched to `uint8_t`.